File: app/src/main/java/org/evcs/android/features/profile/wallet/PaymentMethodPresenter.java
package org.evcs.android.features.profile.wallet;

import com.base.networking.retrofit.RetrofitServices;

import org.evcs.android.model.CreditCard;
import org.evcs.android.network.callback.AuthCallback;
import org.evcs.android.network.service.presenter.ServicesPresenter;
import org.evcs.android.util.ErrorUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import okhttp3.ResponseBody;

public class PaymentMethodPresenter<T extends IPaymentMethodView> extends ServicesPresenter<T> {

    private static final String TAG = "PaymentMethodPresenter";

    protected static final String LOCATION_KEY = "location";
    private AuthCallback<List<CreditCard>> mPaymentMethodUrlCallback;

    public PaymentMethodPresenter(T brainTreeFragment, RetrofitServices services) {
        super(brainTreeFragment, services);

        mPaymentMethodUrlCallback = new AuthCallback<List<CreditCard>>(this) {
            @Override
            public void onResponseSuccessful(List<CreditCard> creditCardInformations) {
                getView().onPaymentMethodsReceived(creditCardInformations);
            }

            @Override
            public void onResponseFailed(ResponseBody responseBody, int i) {
                getView().showError(ErrorUtils.getError(responseBody));
            }

            @Override
            public void onCallFailure(Throwable t) {
                runIfViewCreated(new Runnable() {
                    @Override
                    public void run() {
                        getView().onPaymentMethodsNotReceived();
                    }
                });
            }

        };
    }

    public void getCreditCards() {
        mPaymentMethodUrlCallback.onResponseSuccessful(
                new ArrayList<>(Arrays.asList(new CreditCard()))
        );
//        getService(BrainTreeService.class).getPaymentMethods().enqueue(mPaymentMethodUrlCallback);

    }

    public void makeDefaultPaymentMethod(CreditCard item) {
//        getService(BrainTreeService.class).makeDefaultPaymentMethod(item.getToken())
//                .enqueue(mPaymentMethodUrlCallback);
    }

    public void removePaymentMethod(CreditCard item) {
//        getService(BrainTreeService.class).removePaymentMethod(item.getToken())
//                .enqueue(mPaymentMethodUrlCallback);
    }

}

File: app/src/main/java/org/evcs/android/features/profile/wallet/PaymentMethodPresenter.kt
package org.evcs.android.features.profile.wallet

import com.base.networking.retrofit.RetrofitServices
import org.evcs.android.network.service.presenter.ServicesPresenter
import org.evcs.android.network.callback.AuthCallback
import okhttp3.ResponseBody
import org.evcs.android.network.service.PaymentMethodsService
import org.evcs.android.model.PaymentMethod
import org.evcs.android.util.ErrorUtils
import org.evcs.android.util.UserUtils

class PaymentMethodPresenter<T : IPaymentMethodView>(viewInstance: T, services: RetrofitServices?)
    : ServicesPresenter<T>(viewInstance, services) {

    fun getCreditCards() {
        getService(PaymentMethodsService::class.java).paymentMethods.enqueue(object : AuthCallback<List<PaymentMethod>>(this) {
            override fun onResponseSuccessful(creditCardInformations: List<PaymentMethod>) {
                view.onPaymentMethodsReceived(creditCardInformations)
            }

            override fun onResponseFailed(responseBody: ResponseBody, i: Int) {
                view.showError(ErrorUtils.getError(responseBody))
            }

            override fun onCallFailure(t: Throwable) {
                runIfViewCreated(Runnable { view.onPaymentMethodsNotReceived() })
            }
        })
    }

    fun makeDefaultPaymentMethod(item: PaymentMethod) {
        getService(PaymentMethodsService::class.java).setDefaultPaymentMethod(item.id)
            .enqueue(object : AuthCallback<Void?>(this) {
            override fun onResponseSuccessful(response: Void?) {
                val user = UserUtils.getLoggedUser()
                user.defaultPm = item.id
                UserUtils.saveUser(user)
                view.onDefaultPaymentMethodSet(item)
            }

            override fun onResponseFailed(responseBody: ResponseBody, i: Int) {
                view.showError(ErrorUtils.getError(responseBody))
            }

            override fun onCallFailure(t: Throwable) {
                runIfViewCreated(Runnable { view.onPaymentMethodsNotReceived() })
            }
        })
    }

    fun removePaymentMethod(item: PaymentMethod) {
        getService(PaymentMethodsService::class.java).removePaymentMethod(item.id)
            .enqueue(object : AuthCallback<Void?>(this) {
            override fun onResponseSuccessful(response: Void?) {
                view.onPaymentMethodRemoved(item)
            }

            override fun onResponseFailed(responseBody: ResponseBody, i: Int) {
                view.showError(ErrorUtils.getError(responseBody))
            }

            override fun onCallFailure(t: Throwable) {
                runIfViewCreated(Runnable { view.onPaymentMethodsNotReceived() })
            }
        })
    }

}